### PR TITLE
Refactor derham 2026

### LINKED ISSUE #207: Refactor Derham class
This class needs a refactoring with improved type annotations and getting rid of some dicts in favor of classes.

Todos:

- [x] Refactor unit rests, maybe put longer convergence tests with a mark
- [x] Replace some dicts with classes
- [x] Improve type annotations and docstrings
- [x] Add the possibility of describing boundary conditions with a string Literal: periodic, dirichlet, hom_dirichlet, neumann, ...
- [x] Start writing dev tutorials

### PR DESCRIPTION
**Pull request by spossann**
The signature of `class Derham` has changed, it now reads
```
def __init__(
        self,
        grid: TensorProductGrid,
        options: DerhamOptions,
        comm: MPI.Intracomm = None,
        domain: Domain = None,
        verbose=False,
    ):
```

This should make serialization easier. Added `to_dict()` and `from_dict()` methods to `Derham` class. This is tested in the new dev tutorial for lifting boundary conditions. The lifting is defined for each `FEECVariable` separately through the new attribute `lifting_function`.

Moreover, the following has changed internally:

* Renamed `Nel -> num_elements`
* `p -> degree`
* `nq_pr -> nquads_proj`
* `polar_ck -> polar_splines`
* The type of spline basis functions is now set via the argument `bcs`, which is a tuple of length three. It holds the type for each direction: `None` for periodic, or a tuple of length two with either `"free"` or `"dirichlet"` to indicate the type of clamped splines. `bcs` replaces the two arguments `spl_kind` and `dirichlet_bc`.
* `FEMspaces` are stored under `self.V0fem` etc, whereas `VectorSpaces` are stored under `self.V0` etc.
* The attributes of the 1d spline spaces for each direction and component, such as spline descriptors and quadrature grids, are stored under `self.V0splines` etc. and are assembled in the new class `SplineAttributes1D`.
* The polar spline spaces are stored under `self.V0pol` etc.
* The commuting projectors are stored under `self.P0` etc.

Closes #207

### CODE REVIEW

File: src/struphy/io/options.py [hunk 1]
```diff
@@ -237,6 +278,34 @@ def __repr_no_defaults__(self):
...
+            p=dct["p"],
+            bcs=dct["bcs"],
+            lifting_eta1=dct.get("lifting_eta1", None),
+            lifting_eta2=dct.get("lifting_eta2", None),
+            lifting_eta3=dct.get("lifting_eta3", None),
+            nquads=dct["nquads"],
+            nq_pr=dct["nq_pr"],
+            polar_ck=dct["polar_ck"],
+            local_projectors=dct["local_projectors"],
+        )
```
Comment (max-models) [inline src/struphy/io/options.py]: Is it not possible to inherit from `OptionsBase` and use `to_dict` and `from_dict` from there?

File: src/struphy/io/options.py [hunk 2]
```diff
@@ -274,6 +343,22 @@ def __repr_no_defaults__(self):
...
+        }
+        return dct
+
+    @classmethod
+    def from_dict(cls, dct: dict) -> "FieldsBackground":
+        return cls(
+            type=dct["type"],
+            values=dct["values"],
+            variable=dct["variable"],
+        )
```
Comment (max-models) [inline src/struphy/io/options.py]: Should not be needed since this class is a subclass of `OptionsBase`

File: src/struphy/io/options.py [hunk 3]
```diff
@@ -335,3 +420,31 @@ def __repr_no_defaults__(self):
...
+            out_folders=dct["out_folders"],
+            sim_folder=dct["sim_folder"],
+            restart=dct["restart"],
+            max_runtime=dct["max_runtime"],
+            save_step=dct["save_step"],
+            sort_step=dct["sort_step"],
+            num_clones=dct["num_clones"],
+            profiling_activated=dct["profiling_activated"],
+            profiling_trace=dct["profiling_trace"],
+        )
```
Comment (max-models) [inline src/struphy/io/options.py]: Should not be needed since this class is a subclass of `OptionsBase`

File: src/struphy/feec/psydac_derham.py [hunk 4]
```diff
@@ -41,300 +44,264 @@
...
+    vector space, ``V2`` an ``Hdiv`` vector space, and ``V3`` an ``L2``
+    scalar space.
 
     Parameters
     ----------
-    Nel : list[int]
-        Number of elements in each direction.
+    *spaces : TensorFemSpace | VectorFemSpace
+        Exactly four discrete spaces, ordered as ``(V0, V1, V2, V3)``.
+        The expected pattern is scalar, vector, vector, scalar.
```
Comment (max-models) [inline src/struphy/feec/psydac_derham.py]: If exactly four spaces on this form are always required, why not add v0, v1, v2, v3 as arguments?

File: src/struphy/feec/psydac_derham.py [hunk 5]
```diff
@@ -41,300 +44,264 @@
...
+    - optional local-projector grids,
+    - quadrature grids (points, weights, spans, basis values).
+    """
+
+    def __init__(
+        self,
+        femspace: TensorFemSpace | VectorFemSpace,
+        nquads: tuple[int, int, int],
+        nq_pr: tuple[int, int, int],
+        polar_ck: int = -1,
```
Comment (max-models) [inline src/struphy/feec/psydac_derham.py]: Since you are refactoring, maybe we could also rename variables to avoid abbreviations to make the code more clean?

File: src/struphy/io/options.py (hunk 1 above)
Comment (spossann) [reply src/struphy/io/options.py]: Done [de61901](https://github.com/struphy-hub/struphy/pull/212/commits/de61901648f4eafbf02031d4579fb4c0aaee73c0)

File: src/struphy/io/options.py (hunk 2 above)
Comment (spossann) [reply src/struphy/io/options.py]: Done [de61901](https://github.com/struphy-hub/struphy/pull/212/commits/de61901648f4eafbf02031d4579fb4c0aaee73c0)

File: src/struphy/io/options.py (hunk 3 above)
Comment (spossann) [reply src/struphy/io/options.py]: Done [de61901](https://github.com/struphy-hub/struphy/pull/212/commits/de61901648f4eafbf02031d4579fb4c0aaee73c0)

File: src/struphy/feec/psydac_derham.py (hunk 4 above)
Comment (spossann) [reply src/struphy/feec/psydac_derham.py]: Done [7999037](https://github.com/struphy-hub/struphy/pull/212/commits/79990370ad49edbc7068e97be6322357be36e1b0)

Comment (max-models) [reply src/struphy/feec/psydac_derham.py]: Do you prefer uppercase?

Comment (spossann) [reply src/struphy/feec/psydac_derham.py]: For the spaces yes, to be in line with math notation, since the lowercase v is used for velocities in Struphy.

Comment (max-models) [reply src/struphy/feec/psydac_derham.py]: Makes sense!

File: src/struphy/feec/psydac_derham.py (hunk 5 above)
Comment (spossann) [reply src/struphy/feec/psydac_derham.py]: Done in
* [55d1ce5](https://github.com/struphy-hub/struphy/pull/212/commits/55d1ce54b6ce6a94967cfc5434e1f70881252f2e)
* [85658a0](https://github.com/struphy-hub/struphy/pull/212/commits/85658a001dbaabceb69e219c3ebe28afbda8e53d)
* [c449e2d](https://github.com/struphy-hub/struphy/pull/212/commits/c449e2d4a9d6cc119e9760d8fc11fa6d2c3ea909)

Comment (max-models) [reply src/struphy/feec/psydac_derham.py]: Nice!

**Comment by spossann** [comment]
Hi @max-models - I updated the description. It is now pretty much done, albeit from a renaming of `hom_dirichlet` to `dirichlet` that I will do tomorrow. Maybe you can already give some more feedback.

File: src/struphy/feec/tests/test_derham.py [hunk 6]
```diff
@@ -1,73 +1,152 @@
...
+                "with_projectors": True,
+                "with_local_projectors": True,
+                "spl_kind": (True, True, False),
+                "dirichlet_bc": ((False, False), (False, False), (True, False)),
+            },
+            id="with-local-projectors",
+        ),
+    ],
+)
+def test_psydac_derham(num_elements, init_kwargs, expected):
```
Comment (max-models) [inline src/struphy/feec/tests/test_derham.py:98]: Can we test the `to_dict()` and `from_dict()` here as well?

File: src/struphy/feec/psydac_derham.py [hunk 7]
```diff
@@ -803,32 +1358,56 @@ def args_derham(self):
...
+    @classmethod
+    def from_dict(cls, dct) -> "Derham":
+        """Deserialize a Derham configuration from a dictionary.
+        The MPI communicator is set to None.
+        """
+
+        return cls(
+            grid=TensorProductGrid.from_dict(dct["grid"]),
+            options=DerhamOptions.from_dict(dct["options"]),
+            comm=None,
```
Comment (max-models) [inline src/struphy/feec/psydac_derham.py]: Does this mean that if you do`new_derham_cls = Derham.from_dict(derham_cls.to_dict())` then the `comm` which was used in `derham_cls` is lost?

Maybe it would make more sense to not include the `comm` in the dictionary at all, and instead add it as an argument to `from_dict(cls, dct, comm: MPI.Intracomm = None)` instead?

**Review by max-models — COMMENTED** [review]
It looks really nice! But it's not the easiest PR to review 😅 I will check if I find anything to complain about tomorrow.

File: src/struphy/feec/tests/test_derham.py (hunk 6 above)
Comment (spossann) [reply src/struphy/feec/tests/test_derham.py:98]: Done [51cb47b](https://github.com/struphy-hub/struphy/pull/212/commits/51cb47be207cb7751eace6419f1b82ed27722383)

File: src/struphy/feec/psydac_derham.py (hunk 7 above)
Comment (spossann) [reply src/struphy/feec/psydac_derham.py]: Done [51cb47b](https://github.com/struphy-hub/struphy/pull/212/commits/51cb47be207cb7751eace6419f1b82ed27722383)

**Comment by spossann** [comment]
Hi @szekedidavid -  this branch contains a refactoring of the `Derham` class. It has a new setup for the feec boundary conditions, inspired by your comments on how to set the bcs. The branch will be merged soon into devel (you can already check out from it or merge it into your feature branch).

In particular, have a look at the notebook `tutorials/tutorial_for_devs_01_feec.ipynb`. We now have a basis for treating the lifting boundary conditions in a unified way via the `FEECVariable`. We can discuss more details in our next meeting. Let me know if you have questions - cheers!

**Comment by spossann** [comment]
@max-models finally ready :)

File: src/struphy/feec/psydac_derham.py [hunk 8]
```diff
@@ -390,9 +710,113 @@ def __init__(
...
+        V0pol, V1pol, V2pol, V3pol, Vvpol = self._assemble_polar_extraction_operators()
+
+        self._V0pol = V0pol
+        self._V1pol = V1pol
+        self._V2pol = V2pol
+        self._V3pol = V3pol
+        self._Vvpol = Vvpol
+
+        # Homogeneous Dirichlet boundary operators
+        self._boundary_ops = {}
```
Comment (max-models) [inline src/struphy/feec/psydac_derham.py:739]: So as far as I can see, the only remaining dictionaries are

```
self._boundary_ops = {}
self._extraction_ops = {}
self._dofs_extraction_ops = {}
```

I think it makes sense to keep them as they are, but in principle they could also be converted to classes, but there is no need to add extra complexity.

File: src/struphy/feec/psydac_derham.py [hunk 9]
```diff
@@ -1114,6 +1693,92 @@ def _discretize_space(
 
         return Wh
 
+    def _assemble_polar_extraction_operators(
```
Comment (max-models) [inline src/struphy/feec/psydac_derham.py:1696]: Are `_assemble_polar_extraction_operators`, `_assemble_projectors`, and `_assemble_local_projectors` just separated from the `__init__()` method to make the code cleaner?

**Review by max-models — APPROVED** [review]
I think it looks good!

File: src/struphy/feec/psydac_derham.py (hunk 8 above)
Comment (spossann) [reply src/struphy/feec/psydac_derham.py:739]: Yes I would prefer classes here too, but this will be done when refactoring the polar splines.

File: src/struphy/feec/psydac_derham.py (hunk 9 above)
Comment (spossann) [reply src/struphy/feec/psydac_derham.py:1696]: Yes indeed.

**Comment by spossann** [comment]
> I think it looks good!

Thanks a lot for looking through parts of this 👍
